src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -1412,26 +1412,38 @@ private void EmitCallExpression(BoundCall call, UseKind useKind)
                 else if (receiverType.IsVerifierValue())
                 {
                     NamedTypeSymbol methodContainingType = method.ContainingType;
-                    if (methodContainingType.IsVerifierValue() && MayUseCallForStructMethod(method))
+                    if (methodContainingType.IsVerifierValue())
                     {
-                        // NOTE: this should be either a method which overrides some abstract method or 
-                        //       does not override anything (with few exceptions, see MayUseCallForStructMethod); 
-                        //       otherwise we should not use direct 'call' and must use constrained call;
-
-                        // calling a method defined in a value type
-                        Debug.Assert(receiverType == methodContainingType);
-                        // method is defined in the struct itself and is assumed to be mutating.
-                        //PROTOTYPE(readonlyRefs): when readonly structs are implemented, this will have to check "this"
-                        tempOpt = EmitReceiverRef(receiver, AddressKind.Writeable);
-                        callKind = CallKind.Call;
+                        // if method is defined in the struct itself it is assumed to be mutating, unless 
+                        // it is a member of a readonly struct and is not a constructor
+                        var thisAddresskind = methodContainingType.IsReadOnly && method.MethodKind != MethodKind.Constructor ?
+                                                                        AddressKind.ReadOnly :
+                                                                        AddressKind.Writeable;
```

Comment:
Doesn't `methodContainingType` refer to the type of the method being invoked vs. the type of the method we are presently inside of?